### ISSUE TITLE
Is there memory leak?

### ISSUE THREAD

**Issue by applecandyfromsky** [alanxz/SimpleAmqpClient#159]
code:

```cpp
int main(){
	 vector<AmqpClient::Channel::ptr_t> a;
	 for(int i=0;i<300;i++){
		AmqpClient::Channel::ptr_t b = AmqpClient::Channel::Create( "127.0.0.1", 5672, "hahaha", "hahaha");
		b->DeclareExchange("ex_2gate");
				b->DeclareExchange("ex_2app");
		a.push_back(b);
	}
	std::this_thread::sleep_for(std::chrono::seconds(10));
	for(int i=0;i<300;i++)
		a[i].reset();
	a.resize(0);
	std::this_thread::sleep_for(std::chrono::seconds(60));
}
```

I use top and find that after all the connections have been closed, the program still uses much memory.

**Comment by applecandyfromsky** [comment]
%MEM from 2.0 to 1.6.
I expect it should be close to 0.

**Comment by alanxz** [comment]
What version of SimpleAmqpClient and which version of rabbitmq-c are you using?

Could you run it under address-sanitizer or valgrind to determine if its leaking memory and if so maybe where?

> %MEM from 2.0 to 1.6.
I don't understand what you're trying to say here, could you clarify?

**Comment by applecandyfromsky** [comment]
rabbitmq-c:0.8.0
 SimpleAmqpClient :the latest version
I use  valgrind and get this:
==22833== HEAP SUMMARY:
==22833==     in use at exit: 72,704 bytes in 1 blocks
==22833==   total heap usage: 10,811 allocs, 10,810 frees, 161,072,672 bytes allocated
==22833== 
==22833== LEAK SUMMARY:
==22833==    definitely lost: 0 bytes in 0 blocks
==22833==    indirectly lost: 0 bytes in 0 blocks
==22833==      possibly lost: 0 bytes in 0 blocks
==22833==    still reachable: 72,704 bytes in 1 blocks
==22833==         suppressed: 0 bytes in 0 blocks
==22833== Rerun with --leak-check=full to see details of leaked memory
==22833== 
==22833== For counts of detected and suppressed errors, rerun with: -v
==22833== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

So that means there's no leak?

"%MEM from 2.0 to 1.6."
This is when I use top command,I get this:
     
                                                                          %CPU%MEM
   944 root      20   0  333040  45016   8228 S    2.7    2.9   1:03.43 Xorg        
  3499 ccw       20   0  701228  27656  16960 S   2.7   1.8   0:05.11 gnome-t

When the connections are established,the %MEM shows 2.0,and when they are closed,it changes to 1.6.I think it should be 0.
So maybe I misunderstand the meaning of %MEM.
Thanks.

**Comment by alanxz** [comment]
Yes - I do not observe any leaks from valgrind.

Note that most memory allocators (e.g., tcmalloc) do not immediately return all memory freed using `delete` or `free()` back to the OS, this makes it faster when a program frees then immediately mallocs a new block of memory. Additionally programs use memory in other ways, e.g., the program binary itself consumes a certain amount of memory, same with statically allocated variables.
